libcugraph leaks the RMM dependency to downstream

tl;dr: I don't think the current handling of downloading RMM as part of source code
https://github.com/conda-forge/libcugraph-feedstock/blob/0e96fa5a3edf8cfcd10460d31907de422d67a8e2/recipe/meta.yaml#L13-L16
is appropriate. **RMM must be packaged as a Conda-Forge package** so that libcugraph can depend on it, and downstream packages like CuPy can just depend on libcugraph without worrying about RMM.

From https://github.com/conda-forge/cupy-feedstock/pull/123#issuecomment-825432718:
```
    building 'cupy_backends.cuda.libs.cugraph' extension
    /home/conda/feedstock_root/build_artifacts/cupy_1619157077988/_build_env/bin/x86_64-conda-linux-gnu-cc -DNDEBUG -fwrapv -O2 -Wall -Wstrict-prototypes -march=nocona -mtune=haswell -ftree-vectorize -fPIC -fstack-protector-strong -fno-plt -O2 -ffunction-sections -pipe -isystem /home/conda/feedstock_root/build_artifacts/cupy_1619157077988/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placeh/include -fdebug-prefix-map=/home/conda/feedstock_root/build_artifacts/cupy_1619157077988/work=/usr/local/src/conda/cupy-9.0.0 -fdebug-prefix-map=/home/conda/feedstock_root/build_artifacts/cupy_1619157077988/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placeh=/usr/local/src/conda-prefix -I/usr/local/cuda/include -DNDEBUG -D_FORTIFY_SOURCE=2 -O2 -isystem /home/conda/feedstock_root/build_artifacts/cupy_1619157077988/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placeh/include -I/usr/local/cuda/include -fPIC -D_FORCE_INLINES=1 -DCUPY_CUB_VERSION_CODE=100910 -DCUPY_JITIFY_VERSION_CODE=-1 -I/usr/local/cuda/include/cub -I/tmp/pip-req-build-f19x1qnn/install/../cupy/_core/include -I/usr/local/cuda/include -I/home/conda/feedstock_root/build_artifacts/cupy_1619157077988/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placeh/include -I/home/conda/feedstock_root/build_artifacts/cupy_1619157077988/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placeh/include/cugraph -I/home/conda/feedstock_root/build_artifacts/cupy_1619157077988/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placeh/include/python3.6m -c cupy_backends/cuda/libs/cugraph.cpp -o build/temp.linux-x86_64-3.6/cupy_backends/cuda/libs/cugraph.o
    cc1plus: warning: command line option '-Wstrict-prototypes' is valid for C/ObjC but not for C++
    In file included from /home/conda/feedstock_root/build_artifacts/cupy_1619157077988/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placeh/include/cugraph/algorithms.hpp:18:0,
                     from cupy_backends/cuda/libs/../../cupy_cugraph.h:12,
                     from cupy_backends/cuda/libs/cugraph.cpp:677:
    /home/conda/feedstock_root/build_artifacts/cupy_1619157077988/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placeh/include/cugraph/dendrogram.hpp:18:10: fatal error: rmm/device_uvector.hpp: No such file or directory
     #include <rmm/device_uvector.hpp>
              ^~~~~~~~~~~~~~~~~~~~~~~~
    compilation terminated.
```
https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=307796&view=logs&j=7e9c1385-f003-5191-6e56-de8b58facc56&t=e14c8b26-d4de-545b-9aea-e32da246fb0e&l=4922

It shows RMM headers must be present when building applications depending on libcugraph. But, given that RMM must be pinned at the same version as libcugraph, it is impossible for downstream packages to handle this pinning, and instead libcugraph should do this via `pin_compatible`, which in turn requires RMM to be available as a standalone package.

cc: @kkraus14 @jakirkham @rlratzel

--- a/.scripts/run_docker_build.sh
+++ b/.scripts/run_docker_build.sh
@@ -52,11 +52,11 @@ if [ -z "${DOCKER_IMAGE}" ]; then
         echo "WARNING: DOCKER_IMAGE variable not set and shyaml not installed. Trying to parse with coreutils"
         DOCKER_IMAGE=$(cat .ci_support/${CONFIG}.yaml | grep '^docker_image:$' -A 1 | tail -n 1 | cut -b 3-)
         if [ "${DOCKER_IMAGE}" = "" ]; then
-            echo "No docker_image entry found in ${CONFIG}. Falling back to condaforge/linux-anvil-comp7"
-            DOCKER_IMAGE="condaforge/linux-anvil-comp7"
+            echo "No docker_image entry found in ${CONFIG}. Falling back to quay.io/condaforge/linux-anvil-comp7"
+            DOCKER_IMAGE="quay.io/condaforge/linux-anvil-comp7"
         fi
     else
-        DOCKER_IMAGE="$(cat "${FEEDSTOCK_ROOT}/.ci_support/${CONFIG}.yaml" | shyaml get-value docker_image.0 condaforge/linux-anvil-comp7 )"
+        DOCKER_IMAGE="$(cat "${FEEDSTOCK_ROOT}/.ci_support/${CONFIG}.yaml" | shyaml get-value docker_image.0 quay.io/condaforge/linux-anvil-comp7 )"
     fi
 fi
 

--- a/build-locally.py
+++ b/build-locally.py
@@ -7,6 +7,7 @@ import os
 import glob
 import subprocess
 from argparse import ArgumentParser
+import platform
 
 
 def setup_environment(ns):
@@ -16,10 +17,19 @@ def setup_environment(ns):
         os.environ["BUILD_WITH_CONDA_DEBUG"] = "1"
         if ns.output_id:
             os.environ["BUILD_OUTPUT_ID"] = ns.output_id
+    if "MINIFORGE_HOME" not in os.environ:
+        os.environ["MINIFORGE_HOME"] = os.path.join(
+            os.path.dirname(__file__), "miniforge3"
+        )
 
 
 def run_docker_build(ns):
     script = ".scripts/run_docker_build.sh"
+    subprocess.check_call([script])
+
+
+def run_osx_build(ns):
+    script = ".scripts/run_osx_build.sh"
     subprocess.check_call([script])
 
 
@@ -46,10 +56,16 @@ def verify_config(ns):
     else:
         raise ValueError("config " + ns.config + " is not valid")
     # Remove the following, as implemented
-    if not ns.config.startswith("linux"):
+    if ns.config.startswith("win"):
         raise ValueError(
-            f"only Linux configs currently supported, got {ns.config}"
+            f"only Linux/macOS configs currently supported, got {ns.config}"
         )
+    elif ns.config.startswith("osx") and platform.system() == "Darwin":
+        if "OSX_SDK_DIR" not in os.environ:
+            raise RuntimeError(
+                "Need OSX_SDK_DIR env variable set. Run 'export OSX_SDK_DIR=/opt'"
+                "to download the SDK automatically to '/opt/MacOSX<ver>.sdk'"
+            )
 
 
 def main(args=None):
@@ -68,7 +84,12 @@ def main(args=None):
     verify_config(ns)
     setup_environment(ns)
 
-    run_docker_build(ns)
+    if ns.config.startswith("linux") or (
+        ns.config.startswith("osx") and platform.system() == "Linux"
+    ):
+        run_docker_build(ns)
+    elif ns.config.startswith("osx"):
+        run_osx_build(ns)
 
 
 if __name__ == "__main__":

--- a/recipe/build.sh
+++ b/recipe/build.sh
@@ -3,16 +3,11 @@
 # Derived from cugraph build script, as seen here:
 # https://github.com/rapidsai/cugraph/blob/db20b485cfc5399214afcff604b38493f38e83bf/build.sh#L137
 
-# NOTE: it is assumed the RMM header-only sources have been downloaded from the
-# multi-source "source:" section in the meta.yaml file.
-# cmake must be able to find the RMM headers using find_path(). The RMM_ROOT env
-# var is set so RMM_ROOT/include results in a valid dir for cmake to search.
-
 # Set env var DUMP_LOGS_ON_ERROR=0 to suppress
 DUMP_LOGS_ON_ERROR=${DUMP_LOGS_ON_ERROR:=1}
 
 export CUGRAPH_SRC_DIR="${SRC_DIR}/cugraph"
-export RMM_ROOT="${SRC_DIR}/rmm"
+export RMM_ROOT="${PREFIX}/include/rmm"
 export LIBCUGRAPH_BUILD_DIR=${LIBCUGRAPH_BUILD_DIR:=${CUGRAPH_SRC_DIR}/cpp/build}
 export GPU_ARCHS=ALL
 export INSTALL_PREFIX=${PREFIX:=${CONDA_PREFIX}}

--- a/recipe/meta.yaml
+++ b/recipe/meta.yaml
@@ -31,7 +31,7 @@ requirements:
     - nccl
     - libfaiss =*=*_cuda
     - spdlog =1.7.0
-    - rmm ={{ version }}
+    - librmm ={{ version }}
 
 test:
   commands:

--- a/recipe/meta.yaml
+++ b/recipe/meta.yaml
@@ -10,13 +10,8 @@ source:
     sha256: 719a70e76ee9e68a7a26598ced4dc87cb1b1053286851fce8cf8b14bfd7d480b
     folder: cugraph
 
-  # libcugraph depends on the RMM header-only library of the same version.
-  - url: https://github.com/rapidsai/rmm/archive/refs/tags/v{{ version }}.tar.gz
-    sha256: eddd5b35b016d665eb4e8ea2dd31497d913b3e1eb831124e4cb27cba747267a0
-    folder: rmm
-
 build:
-  number: 0
+  number: 1
   skip: true  # [(not linux) or (cuda_compiler_version == "None")]
   run_exports:
     - {{ pin_subpackage(name, max_pin="x.x") }}
@@ -36,6 +31,7 @@ requirements:
     - nccl
     - libfaiss =*=*_cuda
     - spdlog =1.7.0
+    - rmm ={{ version }}
 
 test:
   commands:
@@ -50,10 +46,8 @@ about:
   license: Apache-2.0 AND BSD-3-Clause
   license_file:
     - cugraph/LICENSE
-    - rmm/LICENSE
     - cugraph/cpp/build/gunrock/src/gunrock_ext/LICENSE.TXT
     - cugraph/cpp/build/_deps/raft-src/LICENSE
-    - cugraph/cpp/build/_deps/thrust-src/LICENSE
     - cugraph/cpp/build/_deps/cuco-src/LICENSE
     - cugraph/cpp/build/_deps/cuhornet-src/LICENSE
   summary: 'GPU accelerated graph algorithms'

--- a/recipe/meta.yaml
+++ b/recipe/meta.yaml
@@ -30,7 +30,6 @@ requirements:
     - boost-cpp
     - nccl
     - libfaiss =*=*_cuda
-    - spdlog =1.7.0
     - librmm ={{ version }}
 
 test:

